[RNMobile] RangeCell styling refactor

lukewalczak opened the pull request:
## Description

**NOTE**: That PR is based on `rnmobile/spacer`
Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1448
Ref to gb-mobile: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1511
Change `RangeCell` mobile component styling.

## How has this been tested?

1. Open ios or android app
2. Add `Spacer` component
3. Press settings button
4. Check `RangeCell` styling
## Screenshots <!-- if applicable -->
ios dark mode unfocused | ios light mode unfocused
--- | ---
<img width="463" alt="Screenshot 2019-10-29 at 10 39 29" src="https://user-images.githubusercontent.com/22746080/67757751-ada89880-fa3c-11e9-8c52-e76e56da93e3.png"> | <img width="507" alt="Screenshot 2019-10-29 at 10 41 09" src="https://user-images.githubusercontent.com/22746080/67757775-b600d380-fa3c-11e9-81a1-8f392df9bd47.png">

ios dark mode focused | ios light mode focused
--- | ---
<img width="463" alt="Screenshot 2019-10-29 at 11 12 03" src="https://user-images.githubusercontent.com/22746080/67757897-f3fdf780-fa3c-11e9-80a1-d899a8c8bcea.png"> | <img width="507" alt="Screenshot 2019-10-29 at 10 41 14" src="https://user-images.githubusercontent.com/22746080/67757869-e0529100-fa3c-11e9-9011-0d607572b4c1.png">

android focused | android unfocused
--- | ---
<img width="427" alt="Screenshot 2019-10-29 at 10 44 07" src="https://user-images.githubusercontent.com/22746080/67757956-10019900-fa3d-11e9-9461-f29e0ccc306b.png"> | <img width="427" alt="Screenshot 2019-10-29 at 10 44 02" src="https://user-images.githubusercontent.com/22746080/67757971-16901080-fa3d-11e9-832c-7ec6273849e9.png">

* Reset Button functionality
<img src="https://user-images.githubusercontent.com/22746080/67758032-2e679480-fa3d-11e9-85ab-7e8696548d85.gif" width="300" />



## Types of changes

* add reset button
* move slider to new line
* change input styling

<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

lukewalczak committed:
Revert package-lock.json

lukewalczak committed:
Merge branch 'master' into rnmobile/spacer

lukewalczak committed:
Setting attributes for spacer height

lukewalczak committed:
Correct the condition for setting maximum value

lukewalczak committed:
Merge branch 'master' into rnmobile/spacer

lukewalczak committed:
Small code refactor

lukewalczak committed:
Improve Accessibility in range-cell

lukewalczak committed:
Merge branch 'master' into rnmobile/spacer

lukewalczak committed:
More accessibility improvements

lukewalczak committed:
Small code refactor

lukewalczak committed:
Styling spacer refactor

lukewalczak committed:
Move logic to RangeCell

lukewalczak committed:
Keep Slider along with TextInput within RangeCell

lukewalczak committed:
Small cleanup

lukewalczak committed:
Merge branch 'master' into rnmobile/spacer

lukewalczak committed:
Fix missing binds

lukewalczak committed:
Fix focusing slider on iphoneX when VO is on

lukewalczak committed:
Adjust a11y voice over

lukewalczak committed:
Refactor RangeCell styles

lukewalczak committed:
Refactor pointerEvents when screen reader is on

lukewalczak committed:
Announce current value when finished

lukewalczak committed:
Merge branch 'rnmobile/spacer' into rnmobile/rangecell-styling

lukewalczak committed:
Small cleanup

lukewalczak commented:
@iamthomasbishop please look at it 🙏

lukewalczak committed:
Improve a11y

lukewalczak committed:
Merge branch 'rnmobile/spacer' into rnmobile/rangecell-styling

iamthomasbishop commented:
@lukewalczak Looking good! I have just one thing that I think we might want to address before shipping, and two suggestions which aren't blockers.

* In Dark Mode, can we change the `Reset` button text and the "filled" part of the slider (everything left of the drag handle) to both use [blue 30](https://color-studio.blog) aka `#5198d9`?

Not blockers:

* Is there any way we can move the relative position of the Slider track row up slightly (8 or 9px if my measurements are correct) without messing up other components? I think every Slider instance will have the same. (This isn't a blocker, but would be nice to tighten up spacing.) Visual example:

<img width="1002" alt="Screen Shot 2019-10-29 at 9 40 33 AM" src="https://user-images.githubusercontent.com/1200432/67778126-be531180-fa30-11e9-9956-3bc97d6898ef.png">

* Suggestion: we can probably remove the icon on the setting row

lukewalczak committed:
Fix a11y on iPhoneX

lukewalczak committed:
Merge branch 'rnmobile/spacer' into rnmobile/rangecell-styling

lukewalczak committed:
Refactor after CR

lukewalczak commented:
Updated according to the comments above:

<img width="300" alt="Screenshot 2019-10-30 at 13 27 58" src="https://user-images.githubusercontent.com/22746080/67860789-1405e780-fb1f-11e9-8124-2d9541368d54.png">

lukewalczak committed:
Update info for translators

iamthomasbishop commented:
@lukewalczak Colors looks better, thank you! One thing I just realized: my last comment gave the incorrect value here –

> Is there any way we can move the relative position of the Slider track row up slightly (8 or 9px if my measurements are correct)

It should have been **12px**, not 8-9. The screenshot I provided was accurate. Apologies for the mix-up.

lukewalczak committed:
Merge branch 'rnmobile/spacer' into rnmobile/rangecell-styling

lukewalczak committed:
Correct styles

lukewalczak commented:
Updated one more time: 
<img width="300" alt="Screenshot 2019-10-30 at 16 41 06" src="https://user-images.githubusercontent.com/22746080/67873685-24748d00-fb34-11e9-8dcd-3b0b78b1a4b7.png">

Also we've discovered that slider on android has additional space before and after slider. I found an [issue](https://github.com/react-native-community/react-native-slider/issues/98) for that in community library. Don't know if this is a system behaviour, but proposed workaround within an issue is not satisfying (it requires more changes in our case).

iamthomasbishop commented:
Changes are looking good, and if we need to address the extra spacing separately, that's fine. You've got my blessing to move forward and ship this 👍

pinarol commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -196,22 +204,30 @@ class BottomSheetCell extends Component {
+						{ allowReset && <TouchableOpacity onPress={ allowReset } accessibilityRole={ 'button' }>
+							<Text style={ resetButtonStyle }>{ sprintf(
+								/* translators: %s: reset */
+								__( '%s' ), resetButtonText	) }
```
__ and _x functions are not working with variables, you need to pass static text in them.

Also do we need sprintf here?

lukewalczak replied:
Fixed

pinarol commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -183,6 +189,8 @@ class BottomSheetCell extends Component {
 
 		const iconStyle = getStylesFromColorScheme( styles.icon, styles.iconDark );
+		const resetButtonStyle = getStylesFromColorScheme( styles.resetButton, styles.resetButtonDark );
+		const resetButtonText = Platform.OS === 'ios' ? 'Reset' : 'RESET';
```
We need to localize these texts here

lukewalczak replied:
Fixed

pinarol commented:
I've changed the base branch since spacer is merged, but it seems like we have merge conflicts so I added In Progress label. Please remove the label after solving them.

pinarol commented:
Also, is there a gutenberg-mobile PR? I couldn't find the link.

lukewalczak committed:
Merge branch 'master' into rnmobile/rangecell-styling

lukewalczak committed:
Refactor texts localizing

lukewalczak commented:
Updated description with a link to gb-mobile PR and resolved conflicts.

jbinda committed:
merge master

jbinda commented:
I have also resolve conflicts after merge [RangeControl markup](https://github.com/WordPress/gutenberg/pull/18036)

etoledom commented on packages/components/src/mobile/bottom-sheet/range-cell.native.js:
```diff
@@ -130,49 +134,60 @@ class BottomSheetRangeCell extends Component {
 				accessible={ accessible }
 				onPress={ this.onCellPress }
 				accessibilityLabel={ accessibilityLabel }
+				allowReset={ this.handleReset }
```
It's interesting that the root `Cell` component provides `Reset` option, but the reset functionality has to be handled by the parent component. It feels to me more like a generic button that could be anything. Maybe we can add it as an accessory view on this component, similar how we added the Switch?

Just by curiosity, I added `allowReset` to other controls, it shows the Reset button but it looks a bit off (apart than it does nothing). It doesn't feel to me that it deserves to be on the root Cell component, unless we plan to add reset to more controls? (In which case this might be a good-enough first step). cc @iamthomasbishop 

![IMG_3459](https://user-images.githubusercontent.com/9772967/68311193-7fb4fb00-00b1-11ea-9751-bb591265d1c8.png)

iamthomasbishop replied:
> It doesn't feel to me that it deserves to be on the root Cell component, unless we plan to add reset to more controls? (In which case this might be a good-enough first step)

That's correct, I don't think there are any other planned components at the moment that will need an explicit `reset` button like this.

iamthomasbishop replied:
One more note: if we do end up adding `reset` to other cells/components, we'll want to display them in the same two-row way which provides more space for the `reset` button.

jbinda replied:
Currently we don't have many `settings` in block but maybe it's not bad idea to allow reset single field. Correct me if I'm wrong but it seems now it's possible to manually delete/change the value or clear all settings.

lukewalczak replied:
* Reset button is placed within a `Cell` component, because we want to locate it next to the label (_Height in pixels_) which is defined there actually, otherwise it has to be positioned absolutely.

* Also we can block `allowReset` button option for other controls to avoid the case

etoledom reviewed (commented):
Thank you for these improvements @lukewalczak - they look great! 🎉 

Added a couple of questions (rather than requests) on the code comments.

And I noticed a few small detail testing with bigger text sizes:

- If the system font sizes are set to a bigger size (quite common size), the fields start cropping the last digit for numbers bigger than 1000. This can be super confusing.
- If the text size is set to the accessibility sizes (the super big ones), the layout start to break. This case is a bit more extreme and we don't really have it well covered in the library in general, so maybe it's not that urgent. But we would want in this case the reset button to be vertically aligned.

![range](https://user-images.githubusercontent.com/9772967/68312043-c22b0780-00b2-11ea-89f9-4db67b45e003.png)

jbinda commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -196,22 +204,28 @@ class BottomSheetCell extends Component {
+						{ allowReset && <TouchableOpacity onPress={ allowReset } accessibilityRole={ 'button' }>
+							<Text style={ resetButtonStyle }>{ resetButtonText }
+							</Text>
+						</TouchableOpacity> }
```
you are right `allowReset` on Web is a `boolean` so it should be consistent on mobile. We take care of it

jbinda replied:
fixed !

jbinda commented:
@etoledom thanks for remarks ! We try to discuss `big font size` issue with Luke tomorrow

jbinda committed:
fix allowReset prop

lukewalczak committed:
Patch allowReset

lukewalczak committed:
a11y improvements

lukewalczak commented:
I did some accessibility tweaks (@etoledom @iamthomasbishop please look at it): 
  * made some space for `Reset` button, but then label is truncated
  * `TextInput` width is multiplied by `PixelRatio.getFontScale()` therefore we can display 4 digits
  * by default `TextInput` width is  `40`

<img src="https://user-images.githubusercontent.com/22746080/68394303-e0a20900-016d-11ea-84db-549b1cb46f58.gif" width="300" />

default | medium | the largest
--- | --- | ---
![def](https://user-images.githubusercontent.com/22746080/68395793-baca3380-0170-11ea-9193-f5ae81ae80b4.png) | ![medium](https://user-images.githubusercontent.com/22746080/68395806-c1f14180-0170-11ea-92b3-3dff204278fb.png) | ![large](https://user-images.githubusercontent.com/22746080/68395811-c4ec3200-0170-11ea-99a6-a7b41e2c5a67.png)



_pay no attention to the DONE button, something was wrong with a simulator_

lukewalczak commented on packages/components/src/mobile/bottom-sheet/range-cell.native.js:
```diff
@@ -37,8 +41,23 @@ class BottomSheetRangeCell extends Component {
+		if ( nextAppState === 'active' ) {
+			this.setState( { fontScale: this.getFontScale() } );
+		}
 	}
```
Honestly, I think that it's worth to save that info in redux store, because it can be useful in the future to control other components 🤔

etoledom replied:
It does sound like it would be helpful! But maybe it's too much for this PR.

etoledom reviewed (commented):
It looks so much better on bigger font sizes! Thanks for the update 🙏 

> made some space for Reset button, but then label is truncated

That's why is better to modify the layout to vertical when the text sizes are set to accessibility sizes. This is simple to do on iOS but not sure on RN. Anyway, I think it's good as it is for now 👍 

> TextInput width is multiplied by PixelRatio.getFontScale() therefore we can display 4 digits

Probably, having more than 4 digits is too much of an edge case for this context in particular. I wonder if there are other usages of the Range Control where the numbers could grow a lot bigger.

I think we can revisit if this case becomes a real problem.

From https://github.com/WordPress/gutenberg/pull/18157#discussion_r343600678
> Reset button is placed within a Cell component, because we want to locate it next to the label (Height in pixels) which is defined there actually, otherwise it has to be positioned absolutely.

Makes sense 👍 
I'm still not sure about having a specific `Reset` button that does nothing, and the reset logic has to be implemented on every instance of it. In this case I'd suggest to add it as a generic button (or "box" holder for any component).

But since it's the only component that uses it so far, it might be OK go forward with it and modify it if we need another control with Reset later on. I'm happy with what @pinarol decides.

pinarol commented:
> otherwise it has to be positioned absolutely

> I'm still not sure about having a specific Reset button that does nothing, and the reset logic has to be implemented on every instance of it. In this case I'd suggest to add it as a generic button (or "box" holder for any component).

I agree both. I think we can make this better with a simple tweak. Let's not name it as Reset in BottomSheetCell if it doesn't reset anything inside that component. How about just naming it as "customAction" and injecting the title and event handler from outside? Something like this:

<BottomSheetCell
    ...
    customAction={title:_('Reset'), onAction:{this.onReset}}
/>

`"box" holder for any component` seems too upfront design to me at this point and without knowing the real use cases it can fail to meet future needs.

lukewalczak commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -196,22 +204,28 @@ class BottomSheetCell extends Component {
+									<View style={ platformStyles.labelIconSeparator } />
+								</View>
+							) }
+							<Text numberOfLines={ 1 } style={ [ defaultLabelStyle, labelStyle ] }>
```
@etoledom If we remove `numberOfLines={ 1 }` we can have a not truncated label and it will look like

<img src="https://user-images.githubusercontent.com/22746080/68469880-52845c00-021b-11ea-9d74-dc539101cfd0.png" width="300" />

lukewalczak replied:
I've also checked how it will affect other settings:

before | after
--- | ---
![Screenshot 2019-11-08 at 11 36 33](https://user-images.githubusercontent.com/22746080/68470591-a3e11b00-021c-11ea-8605-b5e6666e4d7a.png) | ![Screenshot 2019-11-08 at 11 36 28](https://user-images.githubusercontent.com/22746080/68470598-a774a200-021c-11ea-8fff-47b7887173b7.png)

etoledom replied:
Looks better than cropping! 
Let's go with it if it does't affect other instances of cells and controls.

etoledom replied:
> I've also checked how it will affect other settings

Again, better than cropping. Looks good to me :)

lukewalczak commented:
> How about just naming it as "customAction" and injecting the title and event handler from outside?

Works for me! Going to adjust it :)

etoledom commented:
> How about just naming it as "customAction" and injecting the title and event handler from outside?

Sounds good! That's what I meant by a `add it as a generic button` 👍

lukewalczak committed:
Refactor custom button

lukewalczak commented:
@etoledom @pinarol I've refactored it according to your suggestion and added:

`customActionButton={ { title: resetButtonText, handler: this.handleReset } }`

pinarol commented:
Works for me! Also I did a quick testing and it is working pretty smoothly, i am very happy that we got rid of the jumpiness.

lukewalczak committed:
Remove numberOfLines from cell label

etoledom reviewed (approved):
Looks and work great 🎉 
Tested on iOS and Android.

Nice job @lukewalczak !

lukewalczak committed:
Merge branch 'master' into rnmobile/rangecell-styling

lukewalczak committed:
Merge branch 'master' into rnmobile/rangecell-styling
